SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialDispatcher.cpp

```diff
@@ -201,3 +172,101 @@ Worker_ComponentId USpatialDispatcher::GetComponentId(Worker_Op* Op) const
+	});
+}
+uint32 USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const UserCallback<Worker_ComponentUpdateOp>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMPONENT_UPDATE, [&](const Worker_Op* Op) {
+		Callback(Op->component_update);
+	});
+}
+
+uint32 USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const UserCallback<Worker_CommandRequestOp>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMMAND_REQUEST, [&](const Worker_Op* Op) {
+		Callback(Op->command_request);
+	});
+}
+
+uint32 USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const UserCallback<Worker_CommandResponseOp>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMMAND_RESPONSE, [&](const Worker_Op* Op) {
+		Callback(Op->command_response);
+	});
+}
+
+
+uint32 USpatialDispatcher::AddGenericOpCallback(Worker_ComponentId ComponentId, Worker_OpType OpType, TFunction<void(const Worker_Op*)> Callback) {
+	check(SpatialConstants::MIN_EXTERNAL_SCHEMA_ID <= ComponentId && ComponentId <= SpatialConstants::MAX_EXTERNAL_SCHEMA_ID);
+	const uint32 NewCallbackId = NextCallbackId++;
+	ComponentOpTypeToCallbackIdMap.FindOrAdd(ComponentId).FindOrAdd(OpType).Add(NewCallbackId);
+	CallbackIdToDataMap.Add(NewCallbackId, UserOpCallbackData{ ComponentId, OpType, Callback });
+	return NewCallbackId;
+}
+
+void USpatialDispatcher::RemoveOpCallback(CallbackId CallbackId)
+{
+	// Find callback ID in map and assert if it does not exist
+	UserOpCallbackData CallbackDataToRemove = CallbackIdToDataMap.FindAndRemoveChecked(CallbackId);
+
+	TMap<Worker_OpType, TArray<uint32>>* OpTypesToCallbackIds = ComponentOpTypeToCallbackIdMap.Find(CallbackDataToRemove.ComponentId);
+	check(OpTypesToCallbackIds);
+	TArray<uint32>* ComponentCallbackIds = OpTypesToCallbackIds->Find(CallbackDataToRemove.OpType);
+	check(ComponentCallbackIds);
+	check(ComponentCallbackIds->Contains(CallbackId));
+
+	// If removing only callback for a component ID / op type, delete map entries as applicable
+	if (OpTypesToCallbackIds->Num() == 1)
+	{
+		if (ComponentOpTypeToCallbackIdMap.Num() == 1)
+		{
+			ComponentOpTypeToCallbackIdMap.Remove(CallbackDataToRemove.ComponentId);
+			return;
+		}
+		OpTypesToCallbackIds->Remove(CallbackDataToRemove.OpType);
+		return;
+	}
+
+	ComponentCallbackIds->Remove(CallbackId);
+}
+
+void USpatialDispatcher::RunUserCallbacks(Worker_ComponentId ComponentId, const Worker_Op* Op)
+{
+	TMap<Worker_OpType, TArray<uint32>>* OpTypeCallbacks = ComponentOpTypeToCallbackIdMap.Find(ComponentId);
+	if (OpTypeCallbacks == nullptr)
+	{
+		return;
+	}
+
+	TArray<uint32>* ComponentCallbackIds = OpTypeCallbacks->Find(static_cast<Worker_OpType>(Op->op_type));
+	if (ComponentCallbackIds == nullptr)
+	{
+		return;
+	}
+
+	for (uint32& CallbackId : *ComponentCallbackIds)
+	{
+		UserOpCallbackData* CallbackData = CallbackIdToDataMap.Find(CallbackId);
+		check(CallbackData);
+		CallbackData->Callback(Op);
```

Comment:
This is still iterating slowly over a map. We could instead have a vector over of structs containing a callback and an ID, for each op type.

Then we iterate over each array and look for the ID to remove them. If we expect this is currently going to be used rarely this might be slightly simpler than having maps of maps of arrays, and also faster.